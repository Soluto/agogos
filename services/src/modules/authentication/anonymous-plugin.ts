import * as http from 'http';
import { FastifyRequest, FastifyInstance } from 'fastify';
import * as fp from 'fastify-plugin';

declare module 'fastify' {
  interface FastifyRequest<
    HttpRequest = http.IncomingMessage,
    Query = DefaultQuery,
    Params = DefaultParams,
    Headers = DefaultHeaders,
    Body = DefaultBody
  > {
    isAnonymousAccess(): boolean | undefined;
    _isAnonymousAccess: boolean | undefined;
  }
}

function isAnonymousAccess(this: FastifyRequest) {
  return this._isAnonymousAccess;
}

export default fp(async function (instance: FastifyInstance) {
  instance.decorateRequest('isAnonymousAccess', isAnonymousAccess);
});
