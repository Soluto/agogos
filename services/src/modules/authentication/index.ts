export { default as authStrategies } from './strategies';
export { default as getSecret } from './get-secret';
export { default as jwtDecoderPlugin } from './jwt-decoder-plugin';
export { default as anonymousPlugin } from './anonymous-plugin';
