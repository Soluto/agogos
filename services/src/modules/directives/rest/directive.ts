import { SchemaDirectiveVisitor } from 'graphql-tools';
import { GraphQLField } from 'graphql';
import { gql } from 'apollo-server-core';
import { RequestContext } from '../../context';
import { RestParams } from './types';

export class RestDirective extends SchemaDirectiveVisitor {
  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    field.resolve = (source, args, context, info) =>
      context.dataSources.rest.doRequest(this.args as RestParams, { source, args, context, info });
  }
}

export const sdl = gql`
  input KeyValue {
    key: String!
    value: String!
    required: Boolean
  }

  directive @rest(
    url: String!
    method: String
    body: String
    bodyArg: String
    query: [KeyValue!]
    headers: [KeyValue!]
    timeoutMs: Int
    notFoundAsNull: Boolean
  ) on FIELD_DEFINITION
`;
