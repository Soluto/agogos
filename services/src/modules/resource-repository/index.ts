export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
