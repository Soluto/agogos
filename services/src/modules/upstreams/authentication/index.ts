export { getAuthHeaders } from './get-auth-headers';
export { ActiveDirectoryAuth } from './active-directory-auth';
