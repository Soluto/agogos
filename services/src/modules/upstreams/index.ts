import { FastifyRequest } from 'fastify';
import * as _ from 'lodash';
import { injectArgs } from '../arguments-injection';
import { ResourceGroup } from '../resource-repository';
import { ActiveDirectoryAuth, getAuthHeaders } from './authentication';

export { ActiveDirectoryAuth };

export type RequestParams = {
  url: URL;
  method?: string;
  headers?: Record<string, string>;
  body?: BodyInit;
  timeout?: number;
};

export async function applyUpstream(
  requestParams: RequestParams,
  resourceGroup: ResourceGroup,
  activeDirectoryAuth: ActiveDirectoryAuth,
  originalRequest?: Pick<FastifyRequest, 'headers'>
): Promise<RequestParams> {
  const upstream = resourceGroup.upstreams.find(
    u => u.host === requestParams.url.host || u.sourceHosts?.includes(requestParams.url.host)
  );

  if (!upstream) return requestParams;

  if (upstream.headers) {
    if (!requestParams.headers) requestParams.headers = {};

    for (const header of upstream.headers) {
      const headerValue = injectArgs(header.value, { request: originalRequest }) as string;
      requestParams.headers[header.name] = headerValue;
    }
  }

  // Authorization headers
  const headers = await getAuthHeaders(
    upstream,
    resourceGroup.upstreamClientCredentials,
    activeDirectoryAuth,
    originalRequest
  );

  // Replace origin
  if (upstream.targetOrigin) {
    const url = requestParams.url;
    requestParams.url = new URL(url.href.replace(url.origin, upstream.targetOrigin));
  }

  return _.merge({}, requestParams, { headers });
}
