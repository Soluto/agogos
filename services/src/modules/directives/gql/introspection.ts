import fetch from 'node-fetch';
import { createHttpLink } from 'apollo-link-http';
import { ObjectTypeDefinitionNode, parse, printSchema, StringValueNode } from 'graphql';
import { introspectSchema } from 'graphql-tools';
import { ApolloError } from 'apollo-server-core';
import { ApolloLink } from 'apollo-link';
import { setContext } from 'apollo-link-context';
import { RetryLink } from 'apollo-link-retry';
import { FastifyRequest } from 'fastify';
import * as _ from 'lodash';
import { ActiveDirectoryAuth, applyUpstream } from '../../upstreams';
import logger from '../../logger';
import { ResourceGroup } from '../../resource-repository';

export interface RemoteSchema {
  url: string;
  schema: string;
}

export async function updateRemoteGqlSchemas(resourceGroup: ResourceGroup, activeDirectoryAuth: ActiveDirectoryAuth) {
  const knownRemoteGqlServers = new Set(resourceGroup.remoteSchemas?.map(rs => rs.url));
  const unusedRemoteSchemas = new Set(knownRemoteGqlServers);
  const urls = new Set<string>();
  for (const { schema } of resourceGroup.schemas) {
    const typeDefs = parse(schema);
    const objectTypeDefinitions = typeDefs.definitions.filter(
      d => d.kind === 'ObjectTypeDefinition'
    ) as ObjectTypeDefinitionNode[];

    for (const { fields } of objectTypeDefinitions) {
      if (!fields) continue;

      for (const field of fields) {
        const gqlDirectives = field.directives?.filter(d => d.name.value === 'gql');
        if (!gqlDirectives) continue;

        for (const gqlDirective of gqlDirectives) {
          if (!gqlDirective.arguments) continue;

          const urlArgument = gqlDirective.arguments!.find(a => a.name.value === 'url');
          if (!urlArgument) continue;
          const url = (urlArgument!.value as StringValueNode).value;

          if (!knownRemoteGqlServers.has(url)) {
            urls.add(url);
          }

          unusedRemoteSchemas.delete(url);
        }
      }
    }

    const gqlIntrospectUrls = Array.from(urls);
    const newSchemas = await Promise.all(
      gqlIntrospectUrls.map(async url => ({
        url,
        schema: await fetchRemoteGqlSchema(url, resourceGroup, activeDirectoryAuth),
      }))
    );
    if (!newSchemas || newSchemas.length === 0) return;

    if (!resourceGroup.remoteSchemas) {
      resourceGroup.remoteSchemas = [];
    }
    resourceGroup.remoteSchemas.push(...newSchemas);

    _.remove(resourceGroup.remoteSchemas, rs => unusedRemoteSchemas.has(rs.url));
  }
}

async function fetchRemoteGqlSchema(
  url: string,
  resourceGroup: ResourceGroup,
  activeDirectoryAuth: ActiveDirectoryAuth
) {
  try {
    const link = ApolloLink.from([
      new RetryLink({
        delay: { max: 1000 },
        attempts: {
          max: 3,
          retryIf(error) {
            logger.warn({ error, url }, 'Failed fetching introspection query');
            return !!error;
          },
        },
      }),
      setContext(async (_, context) => {
        const requestParams = await applyUpstream(
          {
            url: new URL(url),
          },
          resourceGroup,
          activeDirectoryAuth,
          context?.graphqlContext?.request as FastifyRequest
        );
        return { ...requestParams, uri: requestParams.url };
      }),
      createHttpLink({ fetch: fetch as any }),
    ]);

    const remoteSchema = await introspectSchema(link);
    return printSchema(remoteSchema);
  } catch (err) {
    logger.error({ err, url }, `Introspection query to ${url} failed`);
    throw new ApolloError(`Introspection query to ${url} failed`);
  }
}
