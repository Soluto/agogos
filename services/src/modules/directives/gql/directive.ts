import { GraphQLField } from 'graphql';
import { SchemaDirectiveVisitor, delegateToSchema, makeRemoteExecutableSchema } from 'graphql-tools';
import { ApolloError, gql } from 'apollo-server-core';
import { ApolloLink } from 'apollo-link';
import { createHttpLink } from 'apollo-link-http';
import { setContext } from 'apollo-link-context';
import { FastifyRequest } from 'fastify';
import fetch from 'node-fetch';
import { inject } from '../../arguments-injection';
import { RequestContext } from '../../context';
import { applyUpstream } from '../../upstreams';
import { RemoteSchema } from './introspection';

export class GqlDirective extends SchemaDirectiveVisitor {
  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const { url, fieldName, arguments: gqlArgs, operationType: operationTypeParam, timeoutMs } = this.args;
    const operationType = operationTypeParam?.toLowerCase() ?? 'query';

    const remoteSchemas: RemoteSchema[] = this.context.resourceGroup.remoteSchemas;
    const remoteSchemaResource = remoteSchemas?.find(rs => rs.url === url)?.schema;
    if (!remoteSchemaResource) {
      throw new ApolloError(`Remote schema for ${url} not found in resource group`);
    }

    const link = ApolloLink.from([
      setContext(async (_, context) => {
        const requestParams = await applyUpstream(
          {
            url: new URL(url),
          },
          this.context.resourceGroup,
          context?.graphqlContext?.activeDirectoryAuth,
          context?.graphqlContext?.request as FastifyRequest
        );
        return { ...requestParams, uri: requestParams.url };
      }),
      createHttpLink({
        fetch: fetch as any,
        fetchOptions: { timeout: timeoutMs ?? 10000 },
      }),
    ]);

    const remoteSchema = makeRemoteExecutableSchema({
      schema: remoteSchemaResource,
      link,
    });

    field.resolve = async (source, args, context, info) => {
      return await delegateToSchema({
        schema: remoteSchema,
        operation: operationType,
        fieldName,
        args: inject(gqlArgs, { source, args, context, info }) as Record<string, unknown>,
        context,
        info,
      });
    };
  }
}

export const sdl = gql`
  enum GraphQLOperationType {
    Query
    Mutation
  }

  directive @gql(
    url: String!
    fieldName: String!
    operationType: GraphQLOperationType
    arguments: JSONObject
    timeoutMs: Int
  ) on FIELD_DEFINITION
`;
