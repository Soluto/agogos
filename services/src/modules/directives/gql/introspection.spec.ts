import { gql } from 'apollo-server-core';
import { concatAST, DocumentNode, graphqlSync, print } from 'graphql';
import { IResolvers, makeExecutableSchema } from 'graphql-tools';
import GraphQLJSON, { GraphQLJSONObject } from 'graphql-type-json';
import * as nock from 'nock';
import { ResourceGroup } from '../../resource-repository';
import { ActiveDirectoryAuth } from '../../upstreams/authentication';
import { sdl as gqlDirectiveSdl, GqlDirective } from './directive';
import { updateRemoteGqlSchemas } from './introspection';

interface RemoteServer {
  schema?: DocumentNode;
  responseStatusCode?: number;
  shouldBeSkipped?: boolean;
}

interface TestCase {
  resourceGroup: Partial<ResourceGroup>;
  localSchema: DocumentNode;
  remoteServers: Record<string, RemoteServer>;
}

const testCases: [string, TestCase][] = [
  [
    '1. No remote graphql servers',
    {
      resourceGroup: {},
      localSchema: gql`
        type Query {
          foo: String!
        }
      `,
      remoteServers: {},
    },
  ],
  [
    '2. New unknown remote graphql server',
    {
      resourceGroup: {},
      localSchema: gql`
        type Query {
          foo: String! @gql(url: "http://remote-gql-test-2/graphql", fieldName: "foo")
        }
      `,
      remoteServers: {
        ['http://remote-gql-test-2/graphql']: {
          schema: gql`
            type Query {
              foo: String
            }
          `,
        },
      },
    },
  ],
  [
    '3. Known remote graphql server',
    {
      resourceGroup: {
        remoteSchemas: [
          {
            schema: print(gql`
              type Query {
                foo: String
              }
            `),
            url: 'http://remote-gql-test-3/graphql',
          },
        ],
      },
      localSchema: gql`
        type Query {
          foo: String! @gql(url: "http://remote-gql-test-3/graphql", fieldName: "foo")
        }
      `,
      remoteServers: {
        ['http://remote-gql-test-3/graphql']: {
          shouldBeSkipped: true,
        },
      },
    },
  ],
  [
    '4. One known and one unknown graphql servers',
    {
      resourceGroup: {
        remoteSchemas: [
          {
            schema: print(gql`
              type Query {
                bar: String
              }
            `),
            url: 'http://remote-gql-test-4-1/graphql',
          },
        ],
      },
      localSchema: gql`
        type Query {
          foo: String! @gql(url: "http://remote-gql-test-4-1/graphql", fieldName: "foo")
          bar: String! @gql(url: "http://remote-gql-test-4-2/graphql", fieldName: "bar")
        }
      `,
      remoteServers: {
        ['http://remote-gql-test-4-1/graphql']: {
          shouldBeSkipped: true,
        },
        ['http://remote-gql-test-4-2/graphql']: {
          schema: gql`
            type Query {
              bar: String
            }
          `,
        },
      },
    },
  ],
  [
    '5. Error from remote graphql server',
    {
      resourceGroup: {},
      localSchema: gql`
        type Query {
          foo: String! @gql(url: "http://remote-gql-test-5/graphql", fieldName: "foo")
        }
      `,
      remoteServers: {
        ['http://remote-gql-test-5/graphql']: {
          responseStatusCode: 500,
        },
      },
    },
  ],
];

const mockResponse = (host: string, remoteSchema?: DocumentNode, responseCode = 200) =>
  nock(host)
    .persist()
    .post('/graphql')
    .reply(responseCode, (_url, body: Record<string, any>) =>
      responseCode === 200
        ? graphqlSync({
            schema: makeExecutableSchema({ typeDefs: remoteSchema! }),
            source: body.query,
            variableValues: body.variables,
            operationName: body.operationName,
          })
        : 'Error'
    );

describe('Introspection', () => {
  const scalarsSdl = gql`
    scalar JSON
    scalar JSONObject
  `;

  const scalarsResolvers: IResolvers = {
    JSON: GraphQLJSON,
    JSONObject: GraphQLJSONObject,
  };

  const defaultResourceGroup: ResourceGroup = {
    schemas: [],
    policies: [],
    upstreams: [],
    upstreamClientCredentials: [],
    remoteSchemas: [],
  };

  afterEach(() => {
    nock.cleanAll();
  });

  test.each(testCases)('%s', async (_, { resourceGroup, localSchema, remoteServers }) => {
    const httpCallMocks: [nock.Scope, boolean][] = Object.entries(
      remoteServers
    ).map(([url, { schema, responseStatusCode, shouldBeSkipped }]: [string, RemoteServer]) => [
      mockResponse(new URL(url).origin, schema, responseStatusCode),
      !shouldBeSkipped,
    ]);

    const schema = makeExecutableSchema({
      typeDefs: concatAST([localSchema, gqlDirectiveSdl, scalarsSdl]),
      resolvers: scalarsResolvers,
      schemaDirectives: {
        gql: GqlDirective,
      },
    });

    try {
      await updateRemoteGqlSchemas(schema, { ...defaultResourceGroup, ...resourceGroup }, new ActiveDirectoryAuth());

      httpCallMocks.forEach(httpCallMock => expect(httpCallMock[0].isDone()).toBe(httpCallMock[1]));

      expect(resourceGroup).toMatchSnapshot();
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });
});
