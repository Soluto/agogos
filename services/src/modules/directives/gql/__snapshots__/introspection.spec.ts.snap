// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Introspection 1. No remote graphql servers 1`] = `
Object {
  "policies": Array [],
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Introspection 2. New unknown remote graphql server 1`] = `
Object {
  "policies": Array [],
  "remoteSchemas": Array [
    Object {
      "schema": "type Query {
  foo: String
}
",
      "url": "http://remote-gql-test-2/graphql",
    },
  ],
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Introspection 3. Known remote graphql server 1`] = `
Object {
  "policies": Array [],
  "remoteSchemas": Array [
    Object {
      "schema": "type Query {
  foo: String
}
",
      "url": "http://remote-gql-test-3/graphql",
    },
  ],
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Introspection 4. One known and one unknown graphql servers 1`] = `
Object {
  "policies": Array [],
  "remoteSchemas": Array [
    Object {
      "schema": "type Query {
  bar: String
}
",
      "url": "http://remote-gql-test-4-1/graphql",
    },
    Object {
      "schema": "type Query {
  bar: String
}
",
      "url": "http://remote-gql-test-4-2/graphql",
    },
  ],
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Introspection 5. Error from remote graphql server 1`] = `[Error: Introspection query to http://remote-gql-test-5/graphql failed]`;
